### PR TITLE
Enable workloads to apply crypto functions on inside-out API data

### PR DESCRIPTION
**Pull request by manju956**
Simple wallet demo application is implemented to showcase workloads can access file i/o using inside-out API and apply crypto functions like Encryption, Decryption on the data stored in untrusted boundary.

Simple wallet workload supports following transactions.
1. deposit <account_identifier> <amount> <encryption_key_hex>
2. withdraw <account_identifier> <amount> <encryption_key_hex>
3. transfer <account_identifier1> <account_identifier2> <amount> <encryption_key_hex>

All wallet details are stored in ledger file in the encrypted format using encryption key supplied.
Same hex encoded encryption key of length 64 must be supplied for all above supported transactions. This is because Avalon supports 32 bit symmetric encryption.

README.md file has step by step guide to run simple wallet application.

### CODE REVIEW

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+            Log(TCF_LOG_INFO, "ledger contains no account details");
+            return "";
+        }
+    } else {
+        Log(TCF_LOG_ERROR, "Failed to read wallet balance for account %s",
+        accountName.c_str());
+	    return "";
+    }
+
+    // get wallet balance of given acount
```
Comment (danintel) [inline examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp]: s/acount/account/

File: examples/apps/simple_wallet/README.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
+# Simple Wallet Demo Application
+
+This application demostrates basic wallet transactions like deposit, withdraw and transfer
```
Comment (danintel) [inline examples/apps/simple_wallet/README.md]: s/demostrates/demonstrates/

File: examples/apps/simple_wallet/workload/simple_wallet_logic.cpp [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+std::string WithdrawMoney(std::string accountName, std::string amount,
+    std::string encryptionKey) {
+    std::string walletBalance = ReadWalletBalance(accountName, encryptionKey);
+
+    if (walletBalance.empty()) {
+        Log(TCF_LOG_ERROR, "No account %s details found in the ledger: ",
+            accountName.c_str());
+        return "No wallet detials found";
```
Comment (danintel) [inline examples/apps/simple_wallet/workload/simple_wallet_logic.cpp]: s/detials/details/

File: common/sgx/iohandler/crypto_iohandler.h [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include <string>
+
+std::string DecryptData(std::string cipher, std::string key);
+
+std::string EncryptData(std::string msg, std::string key);
```
Comment (danintel) [inline common/sgx/iohandler/crypto_iohandler.h]: Newline missing.

File: examples/apps/simple_wallet/README.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+5. Execute wallet transactions.
+
+Deposit transaction:  
+
+In Terminal 2, Create wallets for Alice and bob  
+
+```
+./generic_client.py --workload_id "simple-wallet" --in_data "deposit alice 100 
+aabbccddaabbccddaabbccddaabbccddaabbccddaabbccddaabbccddaabbccdd"
```
Comment (danintel) [inline examples/apps/simple_wallet/README.md]: In this example, please explain where I get or generate the encryption key used for this example.

File: examples/apps/simple_wallet/README.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+To use:
+
+1.  If needed, update the Ethereum account and direct registry contract
+    information in `docker/Dockerfile.tcf-dev` and
+    `examples/common/python/connectors/tcf_connector.toml`
+2.  Follow instructions in the "Docker-based Build and Execution" section of
+    the [build document](../../../BUILD.md#dockerbuild) through step 4
+3.  Terminal 1 is running the TCF Enclave Manager and Listener with
+    `docker-compose` and Terminal 2 is running the Docker container.
+4.  In Terminal 2 run `cd $TCF_HOME/examples/apps/generic_client`
```
Comment (danintel) [inline examples/apps/simple_wallet/README.md]: Do I need to start a worker process or something for Simple Wallet? Or is that built-in or automatic?

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    std::string amount) {
+    JSON_Value* ledgerVal;
+    if (ledgerData.empty()) {
+        Log(TCF_LOG_INFO, "Creating first wallet in ledger");
+        ledgerVal = json_value_init_object();
+    } else {
+        // Parse account_ledger.json data
+        ledgerVal = json_parse_string(ledgerData.c_str());
+        tcf::error::ThrowIfNull(
+        ledgerVal, "failed to parse wallet ledger. Bad formatted json");
```
Comment (danintel) [inline examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp]: Change "Bad formatted json" to "Badly formatted JSON" (or "Poorly formatted JSON")

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    size_t resultSize = GetMaxIoResultSize();
+    size_t outBufSize = GetMaxFileSize();
+    outBuf.reserve(outBufSize);
+    result.reserve(resultSize);
+
+    uint32_t status = FileRead(fileHandlerId, LEDGER_FILE,
+        (uint8_t *)result.c_str(), resultSize, (uint8_t *)outBuf.c_str(),
+        ledgerDataSize);
+    if (status == 0) {
+        Log(TCF_LOG_INFO, "File Read operation success. File content %s",
```
Comment (danintel) [inline examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp]: change "success" to "successful" (and lines 135 and 178 below)

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.h [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    std::string accountName);
+
+std::string UpdateLedger(std::string ledgerData, std::string accountName,
+    std::string amount);
+
+std::string ReadWalletBalance(std::string accountName,
+    std::string encryptionKey);
+
+std::string UpdateWalletBalance(std::string accountName, std::string amount,
+    std::string encryptionKey);
```
Comment (danintel) [inline examples/apps/simple_wallet/workload/simple_wallet_execute_io.h]: Add missing newline

File: examples/apps/simple_wallet/workload/simple_wallet_logic.cpp [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+        std::string netWalletBalance = std::to_string(std::stoi(walletBalance) -
+            std::stoi(amount));
+        walletBalance = UpdateWalletBalance(accountName, netWalletBalance, encryptionKey);
+        Log(TCF_LOG_DEBUG, "Net balance after withdrawal: %s", netWalletBalance);
+    }
+
+    return accountName + "::" + walletBalance;
+}
+
+std::string ProcessWalletRequest(std::string strIn) {
```
Comment (danintel) [inline examples/apps/simple_wallet/workload/simple_wallet_logic.cpp]: Can you add a check for a bad format (that is, not a positive number) from the user. For example, entering an amount of `abc `instead of a number, such as `50`. Or is this done elsewhere?

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp (hunk 1 above)
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp]: Done

File: examples/apps/simple_wallet/README.md (hunk 2 above)
Comment (manju956) [reply examples/apps/simple_wallet/README.md]: Done

File: examples/apps/simple_wallet/workload/simple_wallet_logic.cpp (hunk 3 above)
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_logic.cpp]: Done

File: common/sgx/iohandler/crypto_iohandler.cpp [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    } catch (tcf::error::Error& e) {
+        Log(TCF_LOG_ERROR, "error::EncryptData - %d - %s\n", e.error_code(),
+            e.what());
+        return "";
+    } catch (...) {
+        Log(TCF_LOG_ERROR, "error::EncryptData - unknown internal error");
+        return "";
+    }
+    return ByteArrayToBase64EncodedString(cipher);
+}
```
[the remark this answers is not in the document]
Comment (manju956) [reply common/sgx/iohandler/crypto_iohandler.cpp]: Done

File: examples/apps/simple_wallet/workload/simple_wallet_plug-in.cpp [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+            out_wo_data.decrypted_data =
+                ByteArray(result_str.begin(), result_str.end());
+        } else {
+            // Create a new entry
+            out_work_order_data.emplace_back(
+                wo_data.index, ByteArray(result_str.begin(), result_str.end()));
+        }
+        i++;
+    }
+}
```
[the remark this answers is not in the document]
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_plug-in.cpp]: Done

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+        Log(TCF_LOG_INFO, "File Write operation success");
+        g_ledgerDetails = ledgerData;
+    }
+    else {
+        Log(TCF_LOG_ERROR, "File Write operation Failed");
+        return "";
+    }
+    return amount;
+}
+
```
[the remark this answers is not in the document]
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp]: Done

File: common/sgx/iohandler/crypto_iohandler.cpp [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+#include "enclave_utils.h"
+
+std::string DecryptData(std::string cipher, std::string key) {
+    ByteArray cipherBytes = Base64EncodedStringToByteArray(cipher);
+    ByteArray keyBytes = tcf::HexStringToBinary(key);
+    ByteArray msg = {};
+    try {
+        msg = tcf::crypto::skenc::DecryptMessage(keyBytes, cipherBytes);
+    } catch (tcf::error::ValueError& e) {
+        Log(TCF_LOG_ERROR, "error::DecryptData - %d - %s", e.error_code(),
```
[the remark this answers is not in the document]
Comment (manju956) [reply common/sgx/iohandler/crypto_iohandler.cpp]: Done

File: common/sgx/iohandler/crypto_iohandler.h (hunk 4 above)
Comment (manju956) [reply common/sgx/iohandler/crypto_iohandler.h]: Done

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp (hunk 7 above)
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp]: Done

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp (hunk 8 above)
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp]: Done

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.h (hunk 9 above)
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_execute_io.h]: Done

File: examples/apps/simple_wallet/README.md (hunk 6 above)
Comment (manju956) [reply examples/apps/simple_wallet/README.md]: No need of starting a new worker process. Its built-in and statically linked to enclave library.

File: examples/apps/simple_wallet/workload/simple_wallet_logic.cpp (hunk 10 above)
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_logic.cpp]: Ok, will add check.

**Comment by manju956** [comment]
> The only issue left is in file `examples/apps/simple_wallet/README.md ` step 5.
> I do not understand where or how the key is generated. Can they be any random string?
> The example shows:
> `./generic_client.py --workload_id "simple-wallet" --in_data "deposit alice 100 aabbccddaabbccddaabbccddaabbccddaabbccddaabbccddaabbccddaabbccdd" `
> Please show how to generate a key, even if it is an arbitrary random string.
> I do not think `aabbccddaabbccdd...` is a good example.

Dan, i am working on providing an "init transaction" for the client to generate encryption key inside enclave and shared to client. This encryption key can then be used by clients for encrypting ledger file for all the subsequent transactions.

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+        reinterpret_cast<char*>(&serialized_response[0]),
+        serialized_response.size());
+    tcf::error::ThrowIf<tcf::error::RuntimeError>(
+        jret != JSONSuccess, "Wallet serialization failed");
+
+    return std::string(serialized_response.begin(), serialized_response.end());
+
+}
+
+// Read ledger file, decrypt content and store the data in ledgerData
```
Comment (manojgop) [inline examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp]: Couldn't find any decrypt operation in this function

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+uint32_t ExecuteFileRead(std::string& ledgerData, size_t ledgerDataSize) {
+    std::string outBuf;
+    std::string result;
+    uint32_t fileHandlerId = GetIoHandlerId("tcf-base-file-io");
+    size_t resultSize = GetMaxIoResultSize();
+    size_t outBufSize = GetMaxFileSize();
+    outBuf.reserve(outBufSize);
```
Comment (manojgop) [inline examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp]: Should outBufSize be set to ledgerDataSize ?

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.h [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include <string>
+
```
Comment (manojgop) [inline examples/apps/simple_wallet/workload/simple_wallet_execute_io.h]: Could you move all these global functions to C++ class

File: examples/apps/simple_wallet/workload/simple_wallet_logic.cpp [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+#include <string.h>
+#include <stdint.h>
+#include <stdarg.h>
+#include <vector>
+
+#include "simple_wallet_logic.h"
+#include "simple_wallet_execute_io.h"
+#include "enclave_utils.h"
+
+bool IsValidNumber(std::string numStr) {
```
Comment (manojgop) [inline examples/apps/simple_wallet/workload/simple_wallet_logic.cpp]: Did you mean IsValidPositiveInteger ? if so rename the function accordingly. Else you could just rename it to IsValidAmount

File: examples/apps/simple_wallet/workload/simple_wallet_logic.cpp [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+        Log(TCF_LOG_INFO, "Wallet Balance before deposit: %s",
+            walletBalance.c_str());
+        std::string netWalletBalance = std::to_string(std::stoi(walletBalance) +
+            std::stoi(amount));
+        Log(TCF_LOG_INFO, "net Wallet Balance after deposit: %s",
+            netWalletBalance.c_str());
+        walletBalance = UpdateWalletBalance(accountName, netWalletBalance,
+            encryptionKey);
+        Log(TCF_LOG_INFO, "Updated Wallet Balance: %s",
+            walletBalance.c_str());
```
Comment (manojgop) [inline examples/apps/simple_wallet/workload/simple_wallet_logic.cpp]: Why do we need to log both netWalletBalance and Updated Wallet Balance ? Are both different ? If not, can we just log one of it.

File: common/sgx/iohandler/trusted_iohandler.cpp [hunk 20]
```diff
@@ -23,6 +23,16 @@
...
+// Maximum size of buffer that is used for file io
+size_t GetMaxFileSize() {
+    return 1024;
+}
+
+// Maximum size of result buffer that is used to store status of io
+size_t GetMaxIoResultSize() {
+    return 128;
+}
+
```
[the remark this answers is not in the document]
Comment (manju956) [reply common/sgx/iohandler/trusted_iohandler.cpp]: Done

File: common/sgx/iohandler/crypto_iohandler.cpp [hunk 21]
```diff
@@ -0,0 +1,68 @@
+/* Copyright 2019 Intel Corporation
```
[the remark this answers is not in the document]
Comment (manju956) [reply common/sgx/iohandler/crypto_iohandler.cpp]: Ok, will rename the file

File: examples/apps/simple_wallet/README.md [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+
+Generic command line client, `generic_client.py` sends an input request
+with transaction type, account identifier(s), amount and encryption key to the worker
+which access ledger which contains all wallet details in encrypted format using inside-out API.
+The encrypted ledger data is then decrypted by the workload by making use of crypto functions.
+Once the balance is updated, the result is encrypted and written to ledger using file i/o.
+
+Below are the syntax for supported wallet transactions.  
+```
+1. deposit <account_identifier> <amount> <encryption_key_in_hex>
```
[the remark this answers is not in the document]
Comment (manju956) [reply examples/apps/simple_wallet/README.md]: I am working on providing init transaction to generate ledger encryption key inside enclave which will be used for any subsequent client transactions. Hence client need not pass the key as part of in_data. I will update the documentation once the implementation is complete.

File: examples/apps/simple_wallet/README.md [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+
+## Using the Generic Command Line Client to execute Simple Wallet operations
+
+Generic command line client, `generic_client.py` sends an input request
+with transaction type, account identifier(s), amount and encryption key to the worker
+which access ledger which contains all wallet details in encrypted format using inside-out API.
+The encrypted ledger data is then decrypted by the workload by making use of crypto functions.
+Once the balance is updated, the result is encrypted and written to ledger using file i/o.
+
+Below are the syntax for supported wallet transactions.
```
[the remark this answers is not in the document]
Comment (manju956) [reply examples/apps/simple_wallet/README.md]: Done

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp [hunk 24]
```diff
@@ -0,0 +1,187 @@
...
+
+#include "simple_wallet_execute_io.h"
+#include "trusted_iohandler.h"
+#include "crypto_iohandler.h"
+
+#define LEDGER_FILE "/tmp/account_ledger.json"
+
+// Placeholder to store all ledger entries. Useful when updating
+// given account in the ledger.
+std::string g_ledgerDetails;
```
[the remark this answers is not in the document]
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp]: I need to maintain ledger state in workload. Is it a good idea to have class for a single global variable?

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp [hunk 25]
```diff
@@ -0,0 +1,187 @@
...
+        Log(TCF_LOG_INFO, "Successfully created wallet ledger file %s",
+            LEDGER_FILE);
+    } else {
+        Log(TCF_LOG_INFO, "Failed to create wallet ledger file %s",
+            LEDGER_FILE);
+    }
+}
+
+// Get wallet balance for given account from json serialized ledger
+std::string GetWalletBalanceFromLedger(std::string ledgerData,
```
[the remark this answers is not in the document]
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp]: Ok

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp [hunk 26]
```diff
@@ -0,0 +1,187 @@
...
+        "failed to parse the work order request, badly formed JSON");
+
+    JSON_Object* ledger_object = json_value_get_object(parsed);
+    tcf::error::ThrowIfNull(ledger_object, "Missing JSON object in ledger");
+    Log(TCF_LOG_INFO, "Getting balance for the account from ledger object: %s",
+        accountName);
+    std::string balance = GetJsonStr(ledger_object, accountName.c_str(),
+        nullptr);
+    Log(TCF_LOG_INFO, "GetWalletBalanceFromLedger::Got balance for account %s",
+        accountName);
```
[the remark this answers is not in the document]
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp]: It depends. In case of Deposit, if account is not present we need to create a new account.
For other transactions, if account is not present error is thrown (already covered).

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp (hunk 15 above)
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp]: right, will update the comment.

File: examples/apps/simple_wallet/workload/simple_wallet_logic.cpp (hunk 18 above)
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_logic.cpp]: Will rename the method to IsValidAmount.

File: examples/apps/simple_wallet/workload/simple_wallet_logic.cpp (hunk 19 above)
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_logic.cpp]: log to print netWalletBalance was added for debugging purpose, will remove it.

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.h (hunk 17 above)
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_execute_io.h]: Ok

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp (hunk 16 above)
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp]: Yes. will change it.

File: common/sgx/iohandler/file_io.cpp [hunk 27]
```diff
@@ -0,0 +1,151 @@
+/* Copyright 2019 Intel Corporation
```
Comment (manojgop) [inline common/sgx/iohandler/file_io.cpp]: Any specific reason why this is a new file ? My understanding is existing files trusted_iohandler.cpp/.h will we renamed to file_io.cpp/.h. If this is unrelated to this PR , better to raise a separate PR which can be merged first before this one.

File: common/sgx/iohandler/file_io.cpp [hunk 28]
```diff
@@ -0,0 +1,151 @@
...
+#include <stdlib.h>
+#include <stdio.h>
+#include <iostream>
+#include <stdint.h>
+
+#include "file_io.h"
+#include "iohandler_enclave.h"
+#include "enclave_utils.h"
+
+// Maximum size of buffer that is used for file io
```
Comment (manojgop) [inline common/sgx/iohandler/file_io.cpp]: Please avoid global functions. Implement functions inside C++ class and namespace.

File: common/sgx/workload/workload_crypto_utils.cpp [hunk 29]
```diff
@@ -0,0 +1,81 @@
...
+#include "crypto.h"
+#include "error.h"
+#include "tcf_error.h"
+#include "types.h"
+#include "utils.h"
+#include "hex_string.h"
+#include "enclave_utils.h"
+#include "workload_crypto_utils.h"
+
+// Create symmetric encryption key and return hex encoded key string
```
Comment (manojgop) [inline common/sgx/workload/workload_crypto_utils.cpp]: Please avoid global functions. Implement functions inside C++ class and namespace.
I would suggest to move the files crypto_utils.cpp/.h inside simple_wallet/workload. The main crypto functionalities are already in  tcf::crypto: module.

File: examples/apps/simple_wallet/workload/simple_wallet_logic.cpp [hunk 30]
```diff
@@ -0,0 +1,173 @@
...
+    int num = 0;
+    try {
+        num = std::stoi(num_str);
+    } catch (...) {
+        return false;
+    }
+    return (num > 0) ? true : false;
+}
+
+// Create account identified by account_name if not created and
```
Comment (manojgop) [inline examples/apps/simple_wallet/workload/simple_wallet_logic.cpp]: Can you avoid global functions. Place functions inside C++ class/namespace

File: common/sgx/iohandler/file_io.cpp (hunk 27 above)
Comment (manju956) [reply common/sgx/iohandler/file_io.cpp]: this is a custom file io handler file. earlier this file was trusted_iohandler.cpp.

File: common/sgx/iohandler/file_io.cpp (hunk 28 above)
Comment (manju956) [reply common/sgx/iohandler/file_io.cpp]: Done

File: common/sgx/workload/workload_crypto_utils.cpp (hunk 29 above)
Comment (manju956) [reply common/sgx/workload/workload_crypto_utils.cpp]: these are generic crypto util files can be utilized from different workloads. Hence, it can't be moved inside simple_wallet/workload module.

File: examples/apps/simple_wallet/workload/simple_wallet_logic.cpp (hunk 30 above)
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_logic.cpp]: Done

File: common/sgx/iohandler/file_io.cpp (hunk 27 above)
Comment (manojgop) [reply common/sgx/iohandler/file_io.cpp]: Use "git mv" command. This change should ideally be in a separate PR

File: common/sgx/workload/workload_crypto_utils.cpp (hunk 29 above)
Comment (manojgop) [reply common/sgx/workload/workload_crypto_utils.cpp]: Generic crypto functions should be kept in crypto module in tc/sgx/common/crypto.

File: common/sgx/iohandler/file_io.cpp (hunk 27 above)
Comment (manju956) [reply common/sgx/iohandler/file_io.cpp]: Yes, but as part of this PR i have to make some modification to this file. Hence i renamed the file in this PR only.

Comment (manojgop) [reply common/sgx/iohandler/file_io.cpp]: It's not clear what the modifications are. The entire file contents is being shown as new changes in this PR

File: common/sgx/workload/workload_crypto_utils.cpp (hunk 29 above)
Comment (manju956) [reply common/sgx/workload/workload_crypto_utils.cpp]: this is kind of wrapper file to make use of crypto functions. I wanted to seperate usage of crypto functions from simple wallet workload, hence i added this wrapper. Actual implementation of Encryption/Decryption can be part of tc/sgx/common/crypto.

File: common/sgx/iohandler/file_io.cpp (hunk 27 above)
Comment (manju956) [reply common/sgx/iohandler/file_io.cpp]: i will try to delete this file and do "git mv". Hopefully it will be clear.

File: common/sgx/workload/workload_crypto_utils.cpp (hunk 29 above)
Comment (manojgop) [reply common/sgx/workload/workload_crypto_utils.cpp]: There is already crypto_utils.cpp in tc/sgx/common/crypto. You can add these functions in same file. Else we will have multiple related crypto functionalities in multiple places.  Also as of now, I see only simple wallet workload making use of these files. So I would suggest either move it to simple_wallet workload or move it to tc/sgx/common/crypto.

Comment (manju956) [reply common/sgx/workload/workload_crypto_utils.cpp]: Ok, i will add these functions to crypto_utils.cpp in tc/sgx/common/crypto

File: common/sgx/iohandler/file_io.cpp (hunk 27 above)
Comment (manju956) [reply common/sgx/iohandler/file_io.cpp]: Since there are modifications after file renaming, they are appearing as new files.

File: examples/apps/simple_wallet/workload/simple_wallet_logic.cpp [hunk 31]
```diff
@@ -0,0 +1,173 @@
...
+// update wallet balance
+std::string SimpleWalletLogic::DepositMoney(SimpleWalletIoExecutor& sw_io,
+    std::string account_name, std::string amount) {
+    std::string wallet_balance = sw_io.ReadWalletBalance(account_name);
+    if (wallet_balance.empty()) {
+        wallet_balance = sw_io.UpdateWalletBalance(account_name, amount);
```
Comment (Ram-srini) [inline examples/apps/simple_wallet/workload/simple_wallet_logic.cpp]: UpdateWalletBalance() returns empty string on error, Please check for error condition in all the places it is called.

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp [hunk 32]
```diff
@@ -0,0 +1,215 @@
...
+std::string SimpleWalletIoExecutor::UpdateLedger(std::string ledger_data,
+    std::string account_name, std::string amount) {
+    JSON_Value* ledger_val;
+    if (ledger_data.empty()) {
+        Log(TCF_LOG_INFO, "Creating first wallet in ledger");
+        ledger_val = json_value_init_object();
+    } else {
+        // Parse account_ledger.json data
+        ledger_val = json_parse_string(ledger_data.c_str());
+        tcf::error::ThrowIfNull(
```
Comment (Ram-srini) [inline examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp]: This function is throwing an exception on error but other functions returns empty string on error. is caller catching this exception?

File: common/sgx/iohandler/file_io.cpp [hunk 33]
```diff
@@ -0,0 +1,152 @@
...
+  result - status of file open operation
+  result_size - Maximum size of the result buffer
+
+*/
+uint32_t FileIoExecutor::FileOpen(uint32_t handler_id, std::string file_name,
+    uint8_t *result, size_t result_size) {
+    uint32_t status;
+    std::string file_operation = "open";
+    std::string command = file_operation + " " + file_name;
+    status = TcfExecuteIoCommand(handler_id, (const uint8_t *)command.c_str(),
```
[the remark this answers is not in the document]
Comment (manju956) [reply common/sgx/iohandler/file_io.cpp]: workload or enclave can't do file check because system calls are not supported in SGX. But file existence check is added in untrusted section of file i/o api implementations in file_io_processor.cpp

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp (hunk 32 above)
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp]: Ok, will handle the exceptions.

File: examples/apps/simple_wallet/workload/simple_wallet_logic.cpp (hunk 31 above)
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_logic.cpp]: right, will handle it.

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp [hunk 34]
```diff
@@ -0,0 +1,215 @@
...
+
+SimpleWalletLedger* SimpleWalletLedger::GetInstance() {
+    if(!SimpleWalletLedger::instance) {
+        SimpleWalletLedger::instance = new SimpleWalletLedger();
+    }
+    return SimpleWalletLedger::instance;
+}
+
+// Create Wallet ledger file to store details of transactions in encrypted format
+void SimpleWalletIoExecutor::CreateWalletLedger() {
```
[the remark this answers is not in the document]
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp]: CreateWalletLedger is called during simple-wallet workload registration to the enclave. If this fails, its as good as registration failure.

File: common/sgx/iohandler/file_io.cpp [hunk 35]
```diff
@@ -0,0 +1,152 @@
...
+#include "enclave_utils.h"
+
+// Returns iohandler id corresponding to iohandler name
+uint32_t FileIoExecutor::GetIoHandlerId(const char* handler_name) {
+    return TcfGetIoHandlerId(handler_name);
+}
+
+// Maximum size of buffer that is used for file io
+size_t FileIoExecutor::GetMaxFileSize() {
+    return 1024;
```
[the remark this answers is not in the document]
Comment (manju956) [reply common/sgx/iohandler/file_io.cpp]: Ok

File: common/sgx/iohandler/file_io.cpp [hunk 36]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+// Maximum size of result buffer that is used to store status of io
+size_t FileIoExecutor::GetMaxIoResultSize() {
+    return 128;
```
[the remark this answers is not in the document]
Comment (manju956) [reply common/sgx/iohandler/file_io.cpp]: Ok

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp [hunk 37]
```diff
@@ -0,0 +1,215 @@
...
+
+// Create encryption key used for encrypting or decrypting
+// ledger having all wallet details
+bool SimpleWalletIoExecutor::CreateLedgerEncryptionKey() {
+    std::string enc_key = tcf::crypto::CreateHexEncodedEncryptionKey();
+    if (enc_key.empty()) {
+        return false;
+    }
+    SimpleWalletLedger* ledger = SimpleWalletLedger::GetInstance();
+    ledger->ledger_encryption_key = enc_key;
```
[the remark this answers is not in the document]
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp]: One options is to persist encryption key by sgx sealing and unsealing during restart scenario.
But we can't ensure replay protection since latest sgx versions doesn't support monotonic counter.

I will raise a separate PR to take care of persisting key on the file.

File: common/sgx/iohandler/file_io.h [hunk 38]
```diff
@@ -0,0 +1,44 @@
...
+
+class FileIoExecutor {
+public:
+    uint32_t GetIoHandlerId(const char* handlerName);
+
+    size_t GetMaxFileSize();
+
+    size_t GetMaxIoResultSize();
+
+    uint32_t FileOpen(uint32_t handler_id, std::string file_name, uint8_t *result,
```
Comment (manojgop) [inline common/sgx/iohandler/file_io.h]: Do we need to pass handler_id and file_name in all class member functions? We can make handler_id and file_name as private members of class and initialize it in constructor of FileIoExecutor

File: examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp (hunk 37 above)
[the remark this answers is not in the document]
Comment (manju956) [reply examples/apps/simple_wallet/workload/simple_wallet_execute_io.cpp]: I have raised below issue to track.
https://github.com/hyperledger/avalon/issues/198
